src/objective/rank_obj.cu

```diff
@@ -96,16 +274,154 @@ struct PairwiseLambdaWeightComputer {
+      dgroup_dcg_ptr_(dgroup_dcg_->data().get()),
+      dindexable_sorted_preds_pos_(new dh::caching_device_vector<uint32_t>(pred_size)),
+      dindexable_sorted_preds_pos_ptr_(dindexable_sorted_preds_pos_->data().get()) {
+    // Sort the predictions first and get the sorted position
+    SegmentSorter<float> segment_prediction_sorter;
+    segment_prediction_sorter.SortItems(dpreds, pred_size, segment_label_sorter.GroupSegments());
+
+    this->CreateIndexableSortedPredictionPositions(segment_prediction_sorter.OriginalPositions());
+
+    // Compute each group's DCG concurrently
+    // Set the values to be the group indices first so that the predicate knows which
+    // group it is dealing with
+    thrust::sequence(dgroup_dcg_->begin(), dgroup_dcg_->end());
+
+    // TODO(sriramch): parallelize across all elements, if possible
+    // Transform each group - the predictate computes the group's DCG
+    thrust::transform(dgroup_dcg_->begin(), dgroup_dcg_->end(),
+                      dgroup_dcg_->begin(),
+                      ComputeGroupDCG(dsorted_labels_, dgroups_));
+  }
+
+  // While computing the weight that needs to be adjusted by this ranking objective, we need
+  // to figure out where positive and negative labels chosen earlier exists, if the group
+  // were to be sorted by its predictions. To accommodate this, we employ the following algorithm.
+  // For a given group, let's assume the following:
+  // labels:        1 5 9 2 4 8 0 7 6 3
+  // predictions:   1 9 0 8 2 7 3 6 5 4
+  // position:      0 1 2 3 4 5 6 7 8 9
+  //
+  // After label sort:
+  // labels:        9 8 7 6 5 4 3 2 1 0
+  // position:      2 5 7 8 1 4 9 3 0 6
+  //
+  // After prediction sort:
+  // predictions:   9 8 7 6 5 4 3 2 1 0
+  // position:      1 3 5 7 8 9 6 4 0 2
+  //
+  // If a sorted label at position 'x' is chosen, then we need to find out where the prediction
+  // for this label 'x' exists, if the group were to be sorted by predictions.
+  // We first take the sorted prediction positions:
+  // position:      1 3 5 7 8 9 6 4 0 2
+  // at indices:    0 1 2 3 4 5 6 7 8 9
+  //
+  // We create a sorted prediction positional array, such that value at position 'x' gives
+  // us the position in the sorted prediction array where its related prediction lies.
+  // dindexable_sorted_preds_pos_ptr_:  8 0 9 1 7 2 6 3 4 5
+  // at indices:                        0 1 2 3 4 5 6 7 8 9
+  // Basically, swap the previous 2 arrays, sort the indices and reorder positions
+  // for an O(1) lookup using the position where the sorted label exists
+  void CreateIndexableSortedPredictionPositions(const uint32_t *dsorted_preds_pos) {
+    int device_id = -1;
+    dh::safe_cuda(cudaGetDevice(&device_id));
+    uint32_t *dindexable_sorted_preds_pos_ptr = dindexable_sorted_preds_pos_ptr_;
+    // Sort the positions (as indices), and group its indices as sorted prediction positions
+    dh::LaunchN(device_id, dindexable_sorted_preds_pos_->size(),
```

Comment:
I think this is just a thrust scatter operation. Prefer thrust for readability. 

Not sure if you have used permutation iterators before, but they are also useful for these types of situations. You can create a 'virtual' permutation on an array without actually moving data around.